File: Example/SortSnapshotDemoObjC/ViewController.m
//
//  ViewController.m
//  SortSnapshotDemoObjC
//
//  Created by Jinwoo Kim on 6/14/21.
//

#import "ViewController.h"
#import "ViewModel.h"

@interface ViewController () <UICollectionViewDelegate>
@property (weak) UICollectionView *collectionView;
@property ViewModel *viewModel;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setAttributes];
    [self configureCollectionView];
    [self configureBarButtonItems];
    [self configureViewModel];
}

- (void)setAttributes {
    self.title = @"SortSnapshotDemoObjC";
}

- (void)configureCollectionView {
    UICollectionViewCompositionalLayout *collectionViewLayout = [[UICollectionViewCompositionalLayout alloc] initWithSectionProvider:[self makeSectionProvider]];
    UICollectionView *collectionView = [[UICollectionView alloc] initWithFrame:self.view.bounds collectionViewLayout:collectionViewLayout];
    self.collectionView = collectionView;
    
    [self.view addSubview:collectionView];
    collectionView.translatesAutoresizingMaskIntoConstraints = NO;
    [NSLayoutConstraint activateConstraints:@[
        [collectionView.topAnchor constraintEqualToAnchor:self.view.topAnchor],
        [collectionView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor],
        [collectionView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor],
        [collectionView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor]
    ]];
    
    [collectionView registerClass:[UICollectionViewListCell class] forCellWithReuseIdentifier:@"cell"];
    [collectionView registerClass:[UICollectionViewListCell class] forSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:@"headerView"];
    collectionView.delegate = self;
}

- (UICollectionViewCompositionalLayoutSectionProvider)makeSectionProvider {
    return ^NSCollectionLayoutSection * _Nullable(NSInteger section, id<NSCollectionLayoutEnvironment> _Nonnull layoutEnvironment){
        UICollectionLayoutListConfiguration *configuration = [[UICollectionLayoutListConfiguration alloc] initWithAppearance:UICollectionLayoutListAppearanceInsetGrouped];
        
        configuration.headerMode = UICollectionLayoutListHeaderModeSupplementary;
        
        return [NSCollectionLayoutSection sectionWithListConfiguration:configuration layoutEnvironment:layoutEnvironment];
    };
}

- (void)configureBarButtonItems {
    __weak typeof(self) weakSelf = self;
    
    UIBarButtonItem *sortByRandomButton = [[UIBarButtonItem alloc] initWithPrimaryAction:[UIAction actionWithHandler:^(__kindof UIAction * _Nonnull action) {
        [weakSelf.viewModel sortByRandom];
    }]];
    
    UIBarButtonItem *sortByDescendingButton = [[UIBarButtonItem alloc] initWithPrimaryAction:[UIAction actionWithHandler:^(__kindof UIAction * _Nonnull action) {
        [weakSelf.viewModel sortByDescending];
    }]];
    
    UIBarButtonItem *sortByAscendingButton = [[UIBarButtonItem alloc] initWithPrimaryAction:[UIAction actionWithHandler:^(__kindof UIAction * _Nonnull action) {
        [weakSelf.viewModel sortByAscending];
    }]];
    
    sortByRandomButton.image = [UIImage systemImageNamed:@"shuffle"];
    sortByDescendingButton.image = [UIImage systemImageNamed:@"arrow.up"];
    sortByAscendingButton.image = [UIImage systemImageNamed:@"arrow.down"];
    
    self.navigationItem.leftBarButtonItems = @[sortByRandomButton];
    self.navigationItem.rightBarButtonItems = @[sortByDescendingButton, sortByAscendingButton];
}

- (void)configureViewModel {
    ViewModel *viewModel = [[ViewModel alloc] initWithDataSource:[self makeDataSource]];
    self.viewModel = viewModel;
    [viewModel configureInitialData];
}

- (DataSource *)makeDataSource {
    DataSource *dataSource = [[DataSource alloc] initWithCollectionView:self.collectionView
                                                           cellProvider:[self makeCellProvider]];
    
    dataSource.supplementaryViewProvider = [self makeSupplementaryViewProvider];
    
    return dataSource;
}

- (UICollectionViewDiffableDataSourceCellProvider)makeCellProvider {
    return ^UICollectionViewCell * _Nullable(UICollectionView * _Nonnull collectionView, NSIndexPath * _Nonnull indexPath, NSNumber *item) {
        UICollectionViewListCell *cell = (UICollectionViewListCell *)[collectionView dequeueReusableCellWithReuseIdentifier:@"cell" forIndexPath:indexPath];
        
        UIListContentConfiguration *configuration = cell.defaultContentConfiguration;
        configuration.text = [NSString stringWithFormat:@"%lu", item.integerValue];
        cell.contentConfiguration = configuration;
        
        return cell;
    };
}

- (UICollectionViewDiffableDataSourceSupplementaryViewProvider)makeSupplementaryViewProvider {
    __weak typeof(self) weakSelf = self;
    return ^UICollectionReusableView * _Nullable(UICollectionView * _Nonnull collectionView, NSString * _Nonnull elementKind, NSIndexPath * _Nonnull indexPath) {
        NSNumber * _Nullable section = [weakSelf.viewModel getSectionItemFromIndexPath:indexPath];
        
        if (section == nil) return nil;
        
        UICollectionViewListCell *headerView = (UICollectionViewListCell *)[collectionView dequeueReusableSupplementaryViewOfKind:elementKind withReuseIdentifier:@"headerView" forIndexPath:indexPath];
        
        UIListContentConfiguration *configuration = headerView.defaultContentConfiguration;
        configuration.text = [NSString stringWithFormat:@"%lu", section.integerValue];
        headerView.contentConfiguration = configuration;
        
        return headerView;
    };
}

#pragma mark UICollectionViewDelegate

- (BOOL)collectionView:(UICollectionView *)collectionView shouldSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    return NO;
}

- (BOOL)collectionView:(UICollectionView *)collectionView shouldHighlightItemAtIndexPath:(NSIndexPath *)indexPath {
    return NO;
}

@end

File: Example/SortSnapshotDemoSwift/ViewController.swift
//
//  ViewController.swift
//  SortSnapshotDemoSwift
//
//  Created by Jinwoo Kim on 6/14/21.
//

import UIKit

class ViewController: UIViewController {
    private weak var collectionView: UICollectionView!
    private var viewModel: ViewModel!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setAttributes()
        configureCollectionView()
        configureBarButtonItems()
        configureViewModel()
    }
    
    private func setAttributes() {
        title = "SortSnapshotDemoSwift"
    }
    
    private func configureCollectionView() {
        let collectionViewLayout: UICollectionViewCompositionalLayout = .init(sectionProvider: makeSectionProvider())
        let collectionView: UICollectionView = .init(frame: view.bounds, collectionViewLayout: collectionViewLayout)
        self.collectionView = collectionView
        
        view.addSubview(collectionView)
        collectionView.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            collectionView.topAnchor.constraint(equalTo: view.topAnchor),
            collectionView.bottomAnchor.constraint(equalTo: view.bottomAnchor),
            collectionView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            collectionView.trailingAnchor.constraint(equalTo: view.trailingAnchor)
        ])
        
        collectionView.register(UICollectionViewListCell.self, forCellWithReuseIdentifier: "cell")
        collectionView.register(UICollectionViewListCell.self, forSupplementaryViewOfKind: UICollectionView.elementKindSectionHeader, withReuseIdentifier: "headerView")
        collectionView.delegate = self
    }
    
    private func makeSectionProvider() -> UICollectionViewCompositionalLayoutSectionProvider {
        return { (section: Int, layoutEnvironment: NSCollectionLayoutEnvironment) -> NSCollectionLayoutSection? in
            var configuration: UICollectionLayoutListConfiguration = .init(appearance: .insetGrouped)
            
            configuration.headerMode = .supplementary
            
            return .list(using: configuration, layoutEnvironment: layoutEnvironment)
        }
    }
    
    private func configureBarButtonItems() {
        let sortByRandomButton: UIBarButtonItem = .init(title: nil,
                                                        image: UIImage(systemName: "shuffle"),
                                                        primaryAction: .init { [weak self] _ in self?.viewModel.sortByRandom() },
                                                        menu: nil)
        
        let sortByDecendingButton: UIBarButtonItem = .init(title: nil,
                                                           image: UIImage(systemName: "arrow.up"),
                                                           primaryAction: .init { [weak self] _ in self?.viewModel.sortByDecending() },
                                                           menu: nil)
        
        let sortByAscendingButton: UIBarButtonItem = .init(title: nil,
                                                           image: UIImage(systemName: "arrow.down"),
                                                           primaryAction: .init { [weak self] _ in self?.viewModel.sortByAscending() },
                                                           menu: nil)
        
        navigationItem.leftBarButtonItems = [sortByRandomButton]
        navigationItem.rightBarButtonItems = [sortByDecendingButton, sortByAscendingButton]
    }
    
    private func configureViewModel() {
        let viewModel: ViewModel = .init(dataSource: makeDataSource())
        self.viewModel = viewModel
        viewModel.configureInitialData()
    }
    
    private func makeDataSource() -> ViewModel.DataSource {
        let dataSource: ViewModel.DataSource = .init(collectionView: collectionView, cellProvider: makeCellProvider())
        
        dataSource.supplementaryViewProvider = makeSupplementaryViewProvider()
        
        return dataSource
    }
    
    private func makeCellProvider() -> ViewModel.DataSource.CellProvider {
        return { (collectionView, indexPath, item) -> UICollectionViewCell? in
            guard let cell: UICollectionViewListCell = collectionView.dequeueReusableCell(withReuseIdentifier: "cell", for: indexPath) as? UICollectionViewListCell else {
                return nil
            }
            
            var configuration: UIListContentConfiguration = cell.defaultContentConfiguration()
            configuration.text = String(item)
            cell.contentConfiguration = configuration
            
            return cell
        }
    }
    
    private func makeSupplementaryViewProvider() -> ViewModel.DataSource.SupplementaryViewProvider {
        return { [weak self] (collectionView, elementKind, indexPath) -> UICollectionReusableView? in
            guard let section: Int = self?.viewModel.getSectionItem(from: indexPath) else {
                return nil
            }
            
            guard let headerView: UICollectionViewListCell = collectionView.dequeueReusableSupplementaryView(ofKind: elementKind, withReuseIdentifier: "headerView", for: indexPath) as? UICollectionViewListCell else {
                return nil
            }
            
            var configuration: UIListContentConfiguration = headerView.defaultContentConfiguration()
            configuration.text = String(section)
            headerView.contentConfiguration = configuration
            
            return headerView
        }
    }
}

extension ViewController: UICollectionViewDelegate {
    func collectionView(_ collectionView: UICollectionView, shouldSelectItemAt indexPath: IndexPath) -> Bool {
        return false
    }
    
    func collectionView(_ collectionView: UICollectionView, shouldHighlightItemAt indexPath: IndexPath) -> Bool {
        return false
    }
}
